### DISCUSSION TITLE
Combining with other web frameworks / ecosystems

### DISCUSSION THREAD

**Discussion by aschrijver** [leptos-rs/leptos#1206] [Q&A]
Leptos offers an attractive and modern web framework, where one can have an all-Rust back-end to front-end techstack. I am evaluating [wasmCloud](https://wasmcloud.com) on the back-end. That also gives kind of a "WebAssembly. Everywhere." paradigm.

What is less clear to me is how I can still leverage existing web frameworks and benefit from their rich ecosystem. I am particularly thinking of React in this case and specifically [React Flow](https://reactflow.dev/) diagramming component.

Say that I'd like to do as much UI-related stuff in Leptos as I can, what are approaches I could take to integrate this component in the UX?

**Comment by aschrijver** [comment]
Starting to answer my own question with some findings on the web:

- There's a SO answer on [Using NPM packages with Rust and WebAssembly](https://stackoverflow.com/questions/75422119/using-npm-packages-with-rust-and-webassembly).
- [`wasm-bindgen`](https://github.com/rustwasm/wasm-bindgen) can be used, but it doesn't look straight-forward, esp. given [large API](https://reactflow.dev/docs/api/react-flow-props/) of React Flow.
  - Another SO question on [How to load a NPM package to wasm-bindgen](https://stackoverflow.com/questions/73490625/how-to-load-a-npm-package-to-wasm-bindgen). Refers to the top SO answer as solution.
- Yew has a [wasm-bindgen section](https://yew.rs/docs/concepts/basic-web-technologies/wasm-bindgen) in the Docs.
- I started a related discussion in the wasm-bindgen repo: [Using NPM modules in Rust web frameworks](https://github.com/rustwasm/wasm-bindgen/discussions/3488)

**Comment by gbj** [comment]
`wasm-bindgen` is the way to integrate JS libraries with Rust compiled to WebAssembly.

Trying to integrate two frameworks (React and Leptos in this case) is always difficult, and especially so with React, which doesn't necessarily play nicely with the DOM.

The most straightforward way to do this, if you *need* to use React for something, would be to create a web component/custom element that wraps the React component, and then simply use it as an HTML element in your Leptos app.

**Reply by aschrijver** [reply]
Thank you! Yes, I received [an answer](https://github.com/rustwasm/wasm-bindgen/discussions/3488#discussioncomment-6208671) on the `wasm-bindgen` discussions confirming, but it would involve hand-coding the binding. They are not focused on generating the bindings for JS/TS components right now.

I also bumped into [WasmEdge](https://wasmedge.org) that can compile JS code into `.wasm` module. Here are the [Docs on Javascript support](https://wasmedge.org/docs/category/develop-wasm-apps-in-javascript). Found via [Running Javascript in WebAssembly with WasmEdge](https://www.secondstate.io/articles/run-javascript-in-webassembly-with-wasmedge/). It supports ES6 / CommonJS / NodeJS modules and involves first doing a build into a single bundled JS file, as it doesn't know about dependencies. But I think that WasmEdge targets the server-side with WASI et al. I don't know enough about wasm world to say if that makes it unusable here.

The custom web component and subsequent use as HTML element in Leptos sounds interesting. But that means having a dual stack Leptos + React, right?

**Reply by aschrijver** [reply]
I formulated a question in WasmEdge discussions:

- https://github.com/WasmEdge/WasmEdge/discussions/2607

**Reply by gbj** [reply]
Maybe I'm just confused because I don't use React and haven't used React Flow. In general, I'd say that integrating code from any two frameworks (React + Vue, React + Angular, React + Solid, React + Svelte, React + Leptos, whatever) is 90% of what's hard; integrating arbitrary JS into Rust/WASM is relatively easy. 

Not sure what you mean when you say "But that means having a dual stack Leptos + React, right?" I thought that's what you were asking how to do (how to integrate a React component into a Leptos app)

**Reply by aschrijver** [reply]
Confusion may be on my side as well, not knowing how all the moving parts fit together :sweat_smile:

With dual stack I was referring to a setup where part of the page would be rendered by React (and make use of React Flow diagramming component) and other sections of the page would be Leptos UI components. And then interact somehow. It wouldn't be an all-Rust experience. Whereas the other option would entail: This is the NPM module; do some compile magic; integrate into Leptos.

**Reply by gbj** [reply]
Ah, I see. Are we talking about client-side rendering, or server side rendering as well?

You definitely do not want to spin up an entire JS engine inside a Rust server to render some React components. You *could* render Leptos components inside WASM loaded by Node or another JS runtime on the server. But that would also be kind of a pain. So if you're talking about server rendering, mixing languages like this is really not worth it IMO. If there's a particular React library that's really essential to what you're doing, and needs to be rendered on the server, I think you lose a *lot* of the benefit of using a Rust framework. 

If we're just talking about client-side rendering then again, even so you're going to end up running both Leptos and React alongside one another. There's no "compile magic" that can transform an imported NPM package into Rust. Importing an NPM package that uses React means spinning up React to render part of the page, no matter what; again, independent on what language or framework the other thing is, unless it's also JS+React.

**Reply by aschrijver** [reply]
Thanks. In this case the React Flow component is the essential one, and that is client-side rendering. The solutions I've found that allow JS to be compiled to Wasm use [QuickJS](https://github.com/second-state/wasmedge-quickjs) inside of the module apparently. I was just looking at [Extism Javascript PDK](https://github.com/extism/js-pdk#using-with-a-bundler) that looks to be able to do this via `esbuild`. The PDK is a fork of `javy` which uses these internals (snapshot from [Wasm I/O](https://youtu.be/RA5tuHzHpeY?t=670) conference vid):

![Javy internals conceptual architecture](https://github.com/leptos-rs/leptos/assets/5111931/afe727c9-8796-4256-98bf-76f12bff8c0d)

I really don't know if I'm looking into feasible pathways for this integration.. the landscape is kinda confusing to me as a newbie :)

**Reply by gbj** [reply]
Oh, okay. Well if it's client-rendered, what you definitely want to do (IMO) is to wrap it in a web component. What this means is that you can then use it directly in your Leptos app as if it were an ordinary HTML element, which it now is!

```view! { cx,
  <div>
    <react-flow prop:data=something></react-flow>
  </div>
}
```

**Reply by aschrijver** [reply]
Interesting. I would have to experiment with this to see how that works in practice. Thank you for your feedback!

PS. There are some Rust + Wasm diagramming projects I found (but they don't come close to features of React Flow):

- [yew-flow](https://github.com/nilaysavant/yew-flow): Looks like the start of a port of React Flow, but no longer maintained (see: [Demo](https://yew-flow.netlify.app/)).
- [egui_node_graph](https://github.com/setzer22/egui_node_graph): based on [egui](https://github.com/emilk/egui) which is compiled to WebAssembly for the browser.
- [D2: Declarative Diagramming](https://d2lang.com/) is also apparently using WebAssembly in the browser. This is not yet available to integrators.

**Reply by Manubi** [reply]
hey @aschrijver can you tell me more about your findings? I am also interested in knowing about a workflow that let's you use some npm packages in rust. Thanks

**Reply by aschrijver** [reply]
I didn't find more straightforward ways and decided that it is probably prudent to wait some more. The wasm ecosystem is growing fast. Instead I am now looking into Tauri + Wasm + TS/React.

**Reply by Manubi** [reply]
Yeah, I also think the most productive way to go forward is to use some established frontend library like react/svelte.
